<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Statistics with R</title>
  <meta name="description" content="Syllabus for the course ‘Statistics with R’">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="Statistics with R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Syllabus for the course ‘Statistics with R’" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Statistics with R" />
  
  <meta name="twitter:description" content="Syllabus for the course ‘Statistics with R’" />
  

<meta name="author" content="Douwe Molenaar">



  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="classifiers.html">
<link rel="next" href="naivebayes.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />







<script language="javascript">
  function toggle(showHideDiv, switchTextDiv) {
  	var ele = document.getElementById(showHideDiv);
	  var text = document.getElementById(switchTextDiv);
	  if(ele.style.display == "block") {
	  	ele.style.display = "none";
	  	text.innerHTML = "Show solution";
	  }
	  else {
	  	ele.style.display = "block";
      text.innerHTML = "Hide solution";
	  }
  }
</script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="styles/style.css" type="text/css" />
<link rel="stylesheet" href="styles/block_elements.css" type="text/css" />
<link rel="stylesheet" href="styles/localadapt.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Colophon</a></li>
<li class="chapter" data-level="1" data-path="studyguide.html"><a href="studyguide.html"><i class="fa fa-check"></i><b>1</b> Study guide</a><ul>
<li class="chapter" data-level="1.1" data-path="studyguide.html"><a href="studyguide.html#introduction"><i class="fa fa-check"></i><b>1.1</b> Introduction</a></li>
<li class="chapter" data-level="1.2" data-path="studyguide.html"><a href="studyguide.html#entry-conditions"><i class="fa fa-check"></i><b>1.2</b> Entry conditions</a></li>
<li class="chapter" data-level="1.3" data-path="studyguide.html"><a href="studyguide.html#goal-of-the-course"><i class="fa fa-check"></i><b>1.3</b> Goal of the course</a><ul>
<li class="chapter" data-level="1.3.1" data-path="studyguide.html"><a href="studyguide.html#this-is-not-a-statistics-course"><i class="fa fa-check"></i><b>1.3.1</b> This is not a statistics course</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="studyguide.html"><a href="studyguide.html#workload"><i class="fa fa-check"></i><b>1.4</b> Workload</a></li>
<li class="chapter" data-level="1.5" data-path="studyguide.html"><a href="studyguide.html#setup-and-content-of-the-course"><i class="fa fa-check"></i><b>1.5</b> Setup and content of the course</a><ul>
<li class="chapter" data-level="1.5.1" data-path="studyguide.html"><a href="studyguide.html#walk-in-hours"><i class="fa fa-check"></i><b>1.5.1</b> Walk-in hours</a></li>
<li class="chapter" data-level="1.5.2" data-path="studyguide.html"><a href="studyguide.html#syllabus"><i class="fa fa-check"></i><b>1.5.2</b> Syllabus</a></li>
<li class="chapter" data-level="1.5.3" data-path="studyguide.html"><a href="studyguide.html#assignments"><i class="fa fa-check"></i><b>1.5.3</b> Assignments</a></li>
<li class="chapter" data-level="1.5.4" data-path="studyguide.html"><a href="studyguide.html#assessment"><i class="fa fa-check"></i><b>1.5.4</b> Assessment</a></li>
<li class="chapter" data-level="1.5.5" data-path="studyguide.html"><a href="studyguide.html#evaluation-of-the-course"><i class="fa fa-check"></i><b>1.5.5</b> Evaluation of the course</a></li>
<li class="chapter" data-level="1.5.6" data-path="studyguide.html"><a href="studyguide.html#course-schedule"><i class="fa fa-check"></i><b>1.5.6</b> Course schedule</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>I Technical background</b></span></li>
<li class="chapter" data-level="2" data-path="quick-start.html"><a href="quick-start.html"><i class="fa fa-check"></i><b>2</b> A quick start</a><ul>
<li class="chapter" data-level="2.1" data-path="quick-start.html"><a href="quick-start.html#what-is-r"><i class="fa fa-check"></i><b>2.1</b> What is R?</a></li>
<li class="chapter" data-level="2.2" data-path="quick-start.html"><a href="quick-start.html#resources-for-learning-r"><i class="fa fa-check"></i><b>2.2</b> Resources for learning R</a></li>
<li class="chapter" data-level="2.3" data-path="quick-start.html"><a href="quick-start.html#other-sources-on-the-use-of-r"><i class="fa fa-check"></i><b>2.3</b> Other sources on the use of R</a></li>
<li class="chapter" data-level="2.4" data-path="quick-start.html"><a href="quick-start.html#installation-of-r"><i class="fa fa-check"></i><b>2.4</b> Installation of R</a></li>
<li class="chapter" data-level="2.5" data-path="quick-start.html"><a href="quick-start.html#starting-r"><i class="fa fa-check"></i><b>2.5</b> Starting R</a></li>
<li class="chapter" data-level="2.6" data-path="quick-start.html"><a href="quick-start.html#obtaining-help"><i class="fa fa-check"></i><b>2.6</b> Obtaining help</a></li>
<li class="chapter" data-level="2.7" data-path="quick-start.html"><a href="quick-start.html#an-r-session"><i class="fa fa-check"></i><b>2.7</b> An R-session</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html"><i class="fa fa-check"></i><b>3</b> Elementary data types and operations</a><ul>
<li class="chapter" data-level="3.1" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#the-atomic-data-types"><i class="fa fa-check"></i><b>3.1</b> The atomic data types</a></li>
<li class="chapter" data-level="3.2" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#the-basic-data-object-classes"><i class="fa fa-check"></i><b>3.2</b> The basic data object classes</a></li>
<li class="chapter" data-level="3.3" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#assigning-names-to-r-objects"><i class="fa fa-check"></i><b>3.3</b> Assigning names to R-objects</a><ul>
<li class="chapter" data-level="3.3.1" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#listing-objects-in-memory"><i class="fa fa-check"></i><b>3.3.1</b> Listing objects in memory</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#arithmetic-with-vectors-arrays-and-data-frames"><i class="fa fa-check"></i><b>3.4</b> Arithmetic with vectors, arrays, and data frames</a></li>
<li class="chapter" data-level="3.5" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#extracting-and-replacing-parts-of-data-objects"><i class="fa fa-check"></i><b>3.5</b> Extracting and replacing parts of data objects</a><ul>
<li class="chapter" data-level="3.5.1" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#extracting-a-single-sub-object-with-the-double-bracket-index-selector"><i class="fa fa-check"></i><b>3.5.1</b> Extracting a single sub-object with the double bracket index selector <code>[[</code></a></li>
<li class="chapter" data-level="3.5.2" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#for-data-frames-and-lists-and-selectors-behave-similarly"><i class="fa fa-check"></i><b>3.5.2</b> For data frames and lists <code>$</code> and <code>[[</code> selectors behave similarly</a></li>
<li class="chapter" data-level="3.5.3" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#complement-syntax"><i class="fa fa-check"></i><b>3.5.3</b> Selecting portions with the single bracket selector <code>[</code></a></li>
<li class="chapter" data-level="3.5.4" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#changing-parts-of-a-vector"><i class="fa fa-check"></i><b>3.5.4</b> Changing parts of a vector</a></li>
<li class="chapter" data-level="3.5.5" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#effect-of-the-selector-operators-on-data-frames"><i class="fa fa-check"></i><b>3.5.5</b> Effect of the selector operators on data frames</a></li>
<li class="chapter" data-level="3.5.6" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#effect-of-the-selector-operators-on-arrays"><i class="fa fa-check"></i><b>3.5.6</b> Effect of the selector operators on arrays</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#removingelements"><i class="fa fa-check"></i><b>3.6</b> Removing elements from objects</a></li>
<li class="chapter" data-level="3.7" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#coercion"><i class="fa fa-check"></i><b>3.7</b> Data class conversion or coercion</a></li>
<li class="chapter" data-level="3.8" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#exercises"><i class="fa fa-check"></i><b>3.8</b> Exercises</a><ul>
<li class="chapter" data-level="" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#selection"><i class="fa fa-check"></i>Selection</a></li>
<li class="chapter" data-level="" data-path="elementary-datatypes.html"><a href="elementary-datatypes.html#more-selection-vectors-and-sequences"><i class="fa fa-check"></i>More selection, vectors and sequences</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="interacting-with-r.html"><a href="interacting-with-r.html"><i class="fa fa-check"></i><b>4</b> Interacting with R</a><ul>
<li class="chapter" data-level="4.1" data-path="interacting-with-r.html"><a href="interacting-with-r.html#controlling-r-from-a-script"><i class="fa fa-check"></i><b>4.1</b> Controlling R from a script</a></li>
<li class="chapter" data-level="4.2" data-path="interacting-with-r.html"><a href="interacting-with-r.html#other-editors"><i class="fa fa-check"></i><b>4.2</b> Other editors</a></li>
<li class="chapter" data-level="4.3" data-path="interacting-with-r.html"><a href="interacting-with-r.html#working-with-packages"><i class="fa fa-check"></i><b>4.3</b> Working with packages</a></li>
<li class="chapter" data-level="4.4" data-path="interacting-with-r.html"><a href="interacting-with-r.html#where-your-packages-are-installed"><i class="fa fa-check"></i><b>4.4</b> Where your packages are installed</a></li>
<li class="chapter" data-level="4.5" data-path="interacting-with-r.html"><a href="interacting-with-r.html#reproducible-research"><i class="fa fa-check"></i><b>4.5</b> Reproducible research</a><ul>
<li class="chapter" data-level="4.5.1" data-path="interacting-with-r.html"><a href="interacting-with-r.html#staying-organized"><i class="fa fa-check"></i><b>4.5.1</b> Staying organized</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="interacting-with-r.html"><a href="interacting-with-r.html#easy-organizing"><i class="fa fa-check"></i><b>4.6</b> Easy organizing with RStudio</a><ul>
<li class="chapter" data-level="" data-path="interacting-with-r.html"><a href="interacting-with-r.html#rmarkdown-documents"><i class="fa fa-check"></i>Rmarkdown documents</a></li>
<li class="chapter" data-level="" data-path="interacting-with-r.html"><a href="interacting-with-r.html#externalizing-code"><i class="fa fa-check"></i>Externalizing code</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="interacting-with-r.html"><a href="interacting-with-r.html#displaying-information-about-your-r-session"><i class="fa fa-check"></i><b>4.7</b> Displaying information about your R session</a></li>
<li class="chapter" data-level="4.8" data-path="interacting-with-r.html"><a href="interacting-with-r.html#citing-r-and-r-packages-in-reports-and-papers"><i class="fa fa-check"></i><b>4.8</b> Citing R and R-packages in reports and papers</a></li>
<li class="chapter" data-level="4.9" data-path="interacting-with-r.html"><a href="interacting-with-r.html#exercise-diversity-of-deep-sea-nematodes"><i class="fa fa-check"></i><b>4.9</b> Exercise: diversity of deep-sea nematodes</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="graphics.html"><a href="graphics.html"><i class="fa fa-check"></i><b>5</b> Graphics</a><ul>
<li class="chapter" data-level="5.1" data-path="graphics.html"><a href="graphics.html#the-basics-of-r-graphics"><i class="fa fa-check"></i><b>5.1</b> The basics of R graphics</a></li>
<li class="chapter" data-level="5.2" data-path="graphics.html"><a href="graphics.html#x-y-plots"><i class="fa fa-check"></i><b>5.2</b> X-Y plots</a></li>
<li class="chapter" data-level="5.3" data-path="graphics.html"><a href="graphics.html#histograms"><i class="fa fa-check"></i><b>5.3</b> Histograms</a></li>
<li class="chapter" data-level="5.4" data-path="graphics.html"><a href="graphics.html#boxplots"><i class="fa fa-check"></i><b>5.4</b> Boxplots</a></li>
<li class="chapter" data-level="5.5" data-path="graphics.html"><a href="graphics.html#images-and-contour-plots"><i class="fa fa-check"></i><b>5.5</b> Images and contour plots</a></li>
<li class="chapter" data-level="5.6" data-path="graphics.html"><a href="graphics.html#plotting-a-mathematical-function"><i class="fa fa-check"></i><b>5.6</b> Plotting a mathematical function</a></li>
<li class="chapter" data-level="5.7" data-path="graphics.html"><a href="graphics.html#multiple-figures-in-one-graphical-device"><i class="fa fa-check"></i><b>5.7</b> Multiple figures in one graphical device</a></li>
<li class="chapter" data-level="5.8" data-path="graphics.html"><a href="graphics.html#saving-graphs"><i class="fa fa-check"></i><b>5.8</b> Saving graphs</a></li>
<li class="chapter" data-level="5.9" data-path="graphics.html"><a href="graphics.html#graphical-systems"><i class="fa fa-check"></i><b>5.9</b> Different graphical systems and packages</a></li>
<li class="chapter" data-level="5.10" data-path="graphics.html"><a href="graphics.html#exercises-1"><i class="fa fa-check"></i><b>5.10</b> Exercises</a><ul>
<li class="chapter" data-level="" data-path="graphics.html"><a href="graphics.html#simple-curves"><i class="fa fa-check"></i>Simple curves</a></li>
<li class="chapter" data-level="" data-path="graphics.html"><a href="graphics.html#human-population-growth"><i class="fa fa-check"></i>Human population growth</a></li>
<li class="chapter" data-level="" data-path="graphics.html"><a href="graphics.html#toxic-ammonia"><i class="fa fa-check"></i>Toxic ammonia</a></li>
<li class="chapter" data-level="" data-path="graphics.html"><a href="graphics.html#the-iris-data-set"><i class="fa fa-check"></i>The iris data set</a></li>
<li><a href="graphics.html#automatic-coercion-by-plot">Automatic coercion by <code>plot()</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="file-io.html"><a href="file-io.html"><i class="fa fa-check"></i><b>6</b> File input and output</a><ul>
<li class="chapter" data-level="6.1" data-path="file-io.html"><a href="file-io.html#defining-the-path-to-a-file-or-directory"><i class="fa fa-check"></i><b>6.1</b> Defining the path to a file or directory</a></li>
<li class="chapter" data-level="6.2" data-path="file-io.html"><a href="file-io.html#the-working-directory"><i class="fa fa-check"></i><b>6.2</b> The “working directory”</a></li>
<li class="chapter" data-level="6.3" data-path="file-io.html"><a href="file-io.html#using-the-read.table-and-write.table-functions"><i class="fa fa-check"></i><b>6.3</b> Using the <code>read.table</code> and <code>write.table</code> functions</a></li>
<li class="chapter" data-level="6.4" data-path="file-io.html"><a href="file-io.html#reading-data-from-a-webserver"><i class="fa fa-check"></i><b>6.4</b> Reading data from a webserver</a></li>
<li class="chapter" data-level="6.5" data-path="file-io.html"><a href="file-io.html#reading-data-from-compressed-files-and-archives"><i class="fa fa-check"></i><b>6.5</b> Reading data from compressed files and archives</a></li>
<li class="chapter" data-level="6.6" data-path="file-io.html"><a href="file-io.html#inputoutput-with-excel-files-and-database-management-systems"><i class="fa fa-check"></i><b>6.6</b> Input/output with Excel files and database management systems</a></li>
<li class="chapter" data-level="6.7" data-path="file-io.html"><a href="file-io.html#exercises-2"><i class="fa fa-check"></i><b>6.7</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="Programming-with-R.html"><a href="Programming-with-R.html"><i class="fa fa-check"></i><b>7</b> Programming with R</a><ul>
<li class="chapter" data-level="7.1" data-path="Programming-with-R.html"><a href="Programming-with-R.html#defining-a-function"><i class="fa fa-check"></i><b>7.1</b> Defining a function</a></li>
<li class="chapter" data-level="7.2" data-path="Programming-with-R.html"><a href="Programming-with-R.html#program-flow-control"><i class="fa fa-check"></i><b>7.2</b> Program flow control</a></li>
<li class="chapter" data-level="7.3" data-path="Programming-with-R.html"><a href="Programming-with-R.html#literature"><i class="fa fa-check"></i><b>7.3</b> Literature</a></li>
<li class="chapter" data-level="7.4" data-path="Programming-with-R.html"><a href="Programming-with-R.html#exercises-3"><i class="fa fa-check"></i><b>7.4</b> Exercises</a><ul>
<li class="chapter" data-level="7.4.1" data-path="Programming-with-R.html"><a href="Programming-with-R.html#programming-loops"><i class="fa fa-check"></i><b>7.4.1</b> Loops</a></li>
<li class="chapter" data-level="" data-path="Programming-with-R.html"><a href="Programming-with-R.html#diversity-of-deep-sea-nematodes"><i class="fa fa-check"></i>Diversity of deep-sea nematodes</a></li>
<li class="chapter" data-level="" data-path="Programming-with-R.html"><a href="Programming-with-R.html#diversity-indices-a-function-optional"><i class="fa fa-check"></i>Diversity indices – a function (optional)</a></li>
<li class="chapter" data-level="" data-path="Programming-with-R.html"><a href="Programming-with-R.html#rarefaction-diversity-optional"><i class="fa fa-check"></i>Rarefaction diversity (optional)</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="Programming-with-R.html"><a href="Programming-with-R.html#add-prog-ex"><i class="fa fa-check"></i><b>7.5</b> Additional programming exercises</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="vectorization.html"><a href="vectorization.html"><i class="fa fa-check"></i><b>8</b> Vectorization</a><ul>
<li class="chapter" data-level="8.1" data-path="vectorization.html"><a href="vectorization.html#the-function-apply"><i class="fa fa-check"></i><b>8.1</b> The function <code>apply()</code></a></li>
<li class="chapter" data-level="8.2" data-path="vectorization.html"><a href="vectorization.html#the-function-tapply"><i class="fa fa-check"></i><b>8.2</b> The function <code id="the-function-tapply">tapply()</code></a></li>
<li class="chapter" data-level="8.3" data-path="vectorization.html"><a href="vectorization.html#the-functions-lapply-and-sapply"><i class="fa fa-check"></i><b>8.3</b> The functions <code>lapply()</code> and <code>sapply()</code></a></li>
<li class="chapter" data-level="8.4" data-path="vectorization.html"><a href="vectorization.html#exercises-4"><i class="fa fa-check"></i><b>8.4</b> Exercises</a><ul>
<li class="chapter" data-level="8.4.1" data-path="vectorization.html"><a href="vectorization.html#permutation-function"><i class="fa fa-check"></i><b>8.4.1</b> A permutation function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="reshaping.html"><a href="reshaping.html"><i class="fa fa-check"></i><b>9</b> Reshaping and manipulating complex data</a><ul>
<li class="chapter" data-level="9.1" data-path="reshaping.html"><a href="reshaping.html#the-tidyr-package"><i class="fa fa-check"></i><b>9.1</b> The <code>tidyr</code> package</a></li>
<li class="chapter" data-level="9.2" data-path="reshaping.html"><a href="reshaping.html#the-dplyr-package"><i class="fa fa-check"></i><b>9.2</b> The <code>dplyr</code> package</a></li>
<li class="chapter" data-level="9.3" data-path="reshaping.html"><a href="reshaping.html#exercises-5"><i class="fa fa-check"></i><b>9.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="formulasyntax.html"><a href="formulasyntax.html"><i class="fa fa-check"></i><b>10</b> Using formula syntax</a><ul>
<li class="chapter" data-level="10.1" data-path="formulasyntax.html"><a href="formulasyntax.html#using-formula-syntax-in-plotting"><i class="fa fa-check"></i><b>10.1</b> Using formula syntax in plotting</a></li>
<li class="chapter" data-level="10.2" data-path="formulasyntax.html"><a href="formulasyntax.html#using-formula-syntax-in-model-definition"><i class="fa fa-check"></i><b>10.2</b> Using formula syntax in model definition</a><ul>
<li class="chapter" data-level="10.2.1" data-path="formulasyntax.html"><a href="formulasyntax.html#data-with-a-continuous-explanatory-variable"><i class="fa fa-check"></i><b>10.2.1</b> Data with a continuous explanatory variable</a></li>
<li class="chapter" data-level="10.2.2" data-path="formulasyntax.html"><a href="formulasyntax.html#data-with-two-discrete-explanatory-variables-factors"><i class="fa fa-check"></i><b>10.2.2</b> Data with two discrete explanatory variables (factors)</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Applications</b></span></li>
<li class="chapter" data-level="11" data-path="distributions.html"><a href="distributions.html"><i class="fa fa-check"></i><b>11</b> The carnival of distributions</a><ul>
<li class="chapter" data-level="11.1" data-path="distributions.html"><a href="distributions.html#distribution-functions"><i class="fa fa-check"></i><b>11.1</b> Distribution functions</a></li>
<li class="chapter" data-level="11.2" data-path="distributions.html"><a href="distributions.html#distribution-functions-in-r"><i class="fa fa-check"></i><b>11.2</b> Distribution functions in R</a></li>
<li class="chapter" data-level="11.3" data-path="distributions.html"><a href="distributions.html#the-exponential-and-poisson-distributions"><i class="fa fa-check"></i><b>11.3</b> The exponential and Poisson distributions</a><ul>
<li class="chapter" data-level="" data-path="distributions.html"><a href="distributions.html#exercises-6"><i class="fa fa-check"></i>Exercises</a></li>
</ul></li>
<li class="chapter" data-level="11.4" data-path="distributions.html"><a href="distributions.html#the-bernoulli-and-binomial-distributions"><i class="fa fa-check"></i><b>11.4</b> The Bernoulli and binomial distributions</a><ul>
<li class="chapter" data-level="" data-path="distributions.html"><a href="distributions.html#exercises-7"><i class="fa fa-check"></i>Exercises</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="distributions.html"><a href="distributions.html#the-normal-and-standard-normal-distribution"><i class="fa fa-check"></i><b>11.5</b> The normal and standard normal distribution</a><ul>
<li class="chapter" data-level="" data-path="distributions.html"><a href="distributions.html#exercise"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="distributions.html"><a href="distributions.html#students-t-distribution"><i class="fa fa-check"></i><b>11.6</b> Student’s t-distribution</a><ul>
<li class="chapter" data-level="" data-path="distributions.html"><a href="distributions.html#exercise-1"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="distributions.html"><a href="distributions.html#the-chi-squared-distribution"><i class="fa fa-check"></i><b>11.7</b> The chi-squared distribution</a><ul>
<li class="chapter" data-level="11.7.1" data-path="distributions.html"><a href="distributions.html#application-in-pearsons-chi-square-goodness-of-fit-test"><i class="fa fa-check"></i><b>11.7.1</b> Application in Pearson’s chi-square goodness of fit test</a></li>
<li class="chapter" data-level="" data-path="distributions.html"><a href="distributions.html#exercise-2"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="11.8" data-path="distributions.html"><a href="distributions.html#the-f-distribution"><i class="fa fa-check"></i><b>11.8</b> The F-distribution</a><ul>
<li class="chapter" data-level="11.8.1" data-path="distributions.html"><a href="distributions.html#analysis-of-variance"><i class="fa fa-check"></i><b>11.8.1</b> Analysis of variance</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="linearmodels.html"><a href="linearmodels.html"><i class="fa fa-check"></i><b>12</b> Linear models and ANOVA</a><ul>
<li class="chapter" data-level="12.1" data-path="linearmodels.html"><a href="linearmodels.html#modeling-with-factor-type-predictor-variables"><i class="fa fa-check"></i><b>12.1</b> Modeling with factor-type predictor variables</a><ul>
<li class="chapter" data-level="12.1.1" data-path="linearmodels.html"><a href="linearmodels.html#alternative-dummy-variable-coding-schemes"><i class="fa fa-check"></i><b>12.1.1</b> Alternative dummy variable coding schemes</a></li>
<li class="chapter" data-level="" data-path="linearmodels.html"><a href="linearmodels.html#exercises-8"><i class="fa fa-check"></i>Exercises</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="linearmodels.html"><a href="linearmodels.html#two-way-anovafactorial-anova"><i class="fa fa-check"></i><b>12.2</b> Two-way ANOVA/factorial ANOVA</a><ul>
<li class="chapter" data-level="12.2.1" data-path="linearmodels.html"><a href="linearmodels.html#unbalanced-data"><i class="fa fa-check"></i><b>12.2.1</b> Unbalanced data</a></li>
<li class="chapter" data-level="" data-path="linearmodels.html"><a href="linearmodels.html#exercise-5"><i class="fa fa-check"></i>Exercise</a></li>
</ul></li>
<li class="chapter" data-level="12.3" data-path="linearmodels.html"><a href="linearmodels.html#combinations-of-numerical-and-discrete-predictors"><i class="fa fa-check"></i><b>12.3</b> Combinations of numerical and discrete predictors</a><ul>
<li class="chapter" data-level="" data-path="linearmodels.html"><a href="linearmodels.html#exercises-9"><i class="fa fa-check"></i>Exercises</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="linearmodels.html"><a href="linearmodels.html#the-connection-between-linear-models-and-anova"><i class="fa fa-check"></i><b>12.4</b> The connection between linear models and ANOVA</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="hubble.html"><a href="hubble.html"><i class="fa fa-check"></i><b>13</b> The age of the universe</a><ul>
<li class="chapter" data-level="13.1" data-path="hubble.html"><a href="hubble.html#exercises-10"><i class="fa fa-check"></i><b>13.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="datafabrication.html"><a href="datafabrication.html"><i class="fa fa-check"></i><b>14</b> A case of data fabrication</a><ul>
<li class="chapter" data-level="14.1" data-path="datafabrication.html"><a href="datafabrication.html#exercises-11"><i class="fa fa-check"></i><b>14.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="amt-carrier.html"><a href="amt-carrier.html"><i class="fa fa-check"></i><b>15</b> A critical evaluation of ammonium transporter kinetics</a><ul>
<li class="chapter" data-level="15.1" data-path="amt-carrier.html"><a href="amt-carrier.html#exercises-12"><i class="fa fa-check"></i><b>15.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="bias-metabolomics.html"><a href="bias-metabolomics.html"><i class="fa fa-check"></i><b>16</b> Bias in metabolomics data</a><ul>
<li class="chapter" data-level="16.1" data-path="bias-metabolomics.html"><a href="bias-metabolomics.html#exercises-13"><i class="fa fa-check"></i><b>16.1</b> Exercises</a><ul>
<li class="chapter" data-level="16.1.1" data-path="bias-metabolomics.html"><a href="bias-metabolomics.html#an-alternative-solution-generalized-additive-modeling-optional"><i class="fa fa-check"></i><b>16.1.1</b> An alternative solution: Generalized additive modeling (optional)</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="17" data-path="nerve-fiber.html"><a href="nerve-fiber.html"><i class="fa fa-check"></i><b>17</b> Correlation between fiber density and episodic memory?</a><ul>
<li class="chapter" data-level="17.1" data-path="nerve-fiber.html"><a href="nerve-fiber.html#exercises-14"><i class="fa fa-check"></i><b>17.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="classifiers.html"><a href="classifiers.html"><i class="fa fa-check"></i><b>18</b> Logistic regression</a><ul>
<li class="chapter" data-level="18.1" data-path="classifiers.html"><a href="classifiers.html#classifiers"><i class="fa fa-check"></i><b>18.1</b> Classifiers</a></li>
<li class="chapter" data-level="18.2" data-path="classifiers.html"><a href="classifiers.html#logistic-regression"><i class="fa fa-check"></i><b>18.2</b> Logistic regression</a><ul>
<li class="chapter" data-level="18.2.1" data-path="classifiers.html"><a href="classifiers.html#logistic-regression-a-case-of-generalized-linear-modeling"><i class="fa fa-check"></i><b>18.2.1</b> Logistic Regression: a case of Generalized Linear Modeling</a></li>
<li class="chapter" data-level="18.2.2" data-path="classifiers.html"><a href="classifiers.html#logistic-regression-in-r"><i class="fa fa-check"></i><b>18.2.2</b> Logistic regression in R</a></li>
<li class="chapter" data-level="18.2.3" data-path="classifiers.html"><a href="classifiers.html#exercises-15"><i class="fa fa-check"></i><b>18.2.3</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="class-dairy-strains.html"><a href="class-dairy-strains.html"><i class="fa fa-check"></i><b>19</b> Classification of dairy bacteria</a><ul>
<li class="chapter" data-level="19.1" data-path="class-dairy-strains.html"><a href="class-dairy-strains.html#exercises-16"><i class="fa fa-check"></i><b>19.1</b> Exercises</a><ul>
<li class="chapter" data-level="19.1.1" data-path="class-dairy-strains.html"><a href="class-dairy-strains.html#description-of-the-data"><i class="fa fa-check"></i><b>19.1.1</b> Description of the data</a></li>
<li class="chapter" data-level="19.1.2" data-path="class-dairy-strains.html"><a href="class-dairy-strains.html#data-conversion"><i class="fa fa-check"></i><b>19.1.2</b> Data conversion</a></li>
<li class="chapter" data-level="19.1.3" data-path="class-dairy-strains.html"><a href="class-dairy-strains.html#data-analysis"><i class="fa fa-check"></i><b>19.1.3</b> Data analysis</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="20" data-path="naivebayes.html"><a href="naivebayes.html"><i class="fa fa-check"></i><b>20</b> Naïve Bayes classifiers</a><ul>
<li class="chapter" data-level="20.1" data-path="naivebayes.html"><a href="naivebayes.html#the-naive-bayes-classifier"><i class="fa fa-check"></i><b>20.1</b> The Naive Bayes classifier</a></li>
<li class="chapter" data-level="20.2" data-path="naivebayes.html"><a href="naivebayes.html#a-crime-scene"><i class="fa fa-check"></i><b>20.2</b> A crime scene</a></li>
<li class="chapter" data-level="20.3" data-path="naivebayes.html"><a href="naivebayes.html#reconstructing-bayes-lawtheorem"><i class="fa fa-check"></i><b>20.3</b> Reconstructing Bayes law/theorem</a></li>
<li class="chapter" data-level="20.4" data-path="naivebayes.html"><a href="naivebayes.html#deciding-on-the-class"><i class="fa fa-check"></i><b>20.4</b> Deciding on the class</a></li>
<li class="chapter" data-level="20.5" data-path="naivebayes.html"><a href="naivebayes.html#continuous-predicting-variables"><i class="fa fa-check"></i><b>20.5</b> Continuous predicting variables</a></li>
<li class="chapter" data-level="20.6" data-path="naivebayes.html"><a href="naivebayes.html#combining-information-from-different-predictor-variables"><i class="fa fa-check"></i><b>20.6</b> Combining information from different predictor variables</a></li>
<li class="chapter" data-level="20.7" data-path="naivebayes.html"><a href="naivebayes.html#exercises-17"><i class="fa fa-check"></i><b>20.7</b> Exercises</a><ul>
<li class="chapter" data-level="20.7.1" data-path="naivebayes.html"><a href="naivebayes.html#predicting-iris-species"><i class="fa fa-check"></i><b>20.7.1</b> Predicting iris species</a></li>
<li class="chapter" data-level="20.7.2" data-path="naivebayes.html"><a href="naivebayes.html#predicting-income"><i class="fa fa-check"></i><b>20.7.2</b> Predicting income</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="21" data-path="resampling-techniques.html"><a href="resampling-techniques.html"><i class="fa fa-check"></i><b>21</b> Resampling techniques</a><ul>
<li class="chapter" data-level="21.1" data-path="resampling-techniques.html"><a href="resampling-techniques.html#bootstrapping"><i class="fa fa-check"></i><b>21.1</b> Bootstrapping</a></li>
<li class="chapter" data-level="21.2" data-path="resampling-techniques.html"><a href="resampling-techniques.html#exercises-18"><i class="fa fa-check"></i><b>21.2</b> Exercises</a><ul>
<li class="chapter" data-level="21.2.1" data-path="resampling-techniques.html"><a href="resampling-techniques.html#medical-patches"><i class="fa fa-check"></i><b>21.2.1</b> Medical patches</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="resampling-techniques.html"><a href="resampling-techniques.html#permutation-test"><i class="fa fa-check"></i><b>21.3</b> Permutation test</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="numerical-differentiation.html"><a href="numerical-differentiation.html"><i class="fa fa-check"></i><b>22</b> Numerical differentiation and smoothing</a><ul>
<li class="chapter" data-level="22.1" data-path="numerical-differentiation.html"><a href="numerical-differentiation.html#exercises-19"><i class="fa fa-check"></i><b>22.1</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="23" data-path="randomnumbers.html"><a href="randomnumbers.html"><i class="fa fa-check"></i><b>23</b> Random numbers</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistics with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="class-dairy-strains" class="section level1">
<h1><span class="header-section-number">CHAPTER 19</span> Classification of dairy bacteria</h1>
<p><strong>(Techniques: hierarchical clustering, decision tree classifier)</strong></p>
<p>This section deals with the data from an experiment by <span class="citation">Bachmann et al. (<a href="#ref-Bachmann2009">2009</a>)</span> in which the diversity of enzyme activities was assessed in different strains of the bacterial species <em>Lactococcus lactis</em>. A few other, independently determined characteristics of these strains are also included. Several strains of this organism are used in dairy fermentations, in particular in the production of cheese. <em>Lactococcus lactis</em> is the main bacterium active in ripening cheeses. During ripening these bacteria convert a small part of the milk proteins and fats to amino acids and volatile compounds. These are an important determinant of the flavour of cheese.</p>
<p>The data is available on the server (<a href="data/strain_diversity">data/strain_diversity</a>) in the file “straindiversity.dif”. The file has a format called “DIF” (Data Interchange Format). DIF is a text-based format designed for spreadsheet-like data. You can find more information about this format on the internet. R has a standard function, <code>read.DIF()</code>, for reading such files.</p>
<div id="exercises-16" class="section level2">
<h2><span class="header-section-number">19.1</span> Exercises</h2>
<ol style="list-style-type: decimal">
<li>Read the data from the server in an R data frame using the <code>read.DIF()</code> function. The file has a header and also contains missing data entries, indicated by the string “nd”. Use the correct settings of the <code>read.DIF()</code> function to automatically convert the header into the column names of the data frame and to set the missing entries to the value <code>NA</code>. Print a summary of the data on screen with the <code>summary()</code> function.
<div>
<a id="BiasMetabHead1" href="javascript:toggle('BiasMetabSol1','BiasMetabHead1');" >Show solution</a>
</div>
<div id="BiasMetabSol1" style="display: none">
<p>Loading the data with the <code>read.DIF()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">f &lt;-<span class="st"> </span><span class="kw">file.path</span>(baseurl, <span class="st">&#39;data&#39;</span>, <span class="st">&#39;strain_diversity&#39;</span>, <span class="st">&#39;straindiversity.dif&#39;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d &lt;-<span class="st"> </span><span class="kw">read.DIF</span>(f, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">na.strings=</span><span class="st">&#39;nd&#39;</span>)</code></pre></div>
</div></li>
</ol>
<div id="description-of-the-data" class="section level3">
<h3><span class="header-section-number">19.1.1</span> Description of the data</h3>
<p>The columns “Strain” and “Other.Strain.Code” contain strains names of the NIZO (former Netherlands Institute for Dairy Research) strain collection and strain codes from other collections and publications. The “Origin” column has two levels, showing whether the strain was collected from a dairy product or was isolated from another source (plant material, for example). The “TaxonomicPosition” column gives the official taxonomic name of the strain. There are three levels in this column. The “Variant” column indicates the subspecies or variant type. There are three levels in this column. The column “Cremoris.Like.Genotype” contains logical values, coded as 1 for TRUE and 0 for FALSE. This characteristic is based on sequence information<a href="#fn14" class="footnoteRef" id="fnref14"><sup>14</sup></a>. Some subspecies of the “lactis” type actually display more similarity to the genotype thought to be specific for the subspecies “cremoris”. They are still called lactis subspecies, because of a set of phenotypic characteristics.</p>
<p>The subsequent ten columns contain activity measurements of five enzymes measured in the strains when the bacteria are grown either on a growth medium called GM17 or on a chemically defined growth medium called CDM. GM17 is a rich medium of largely unknown composition containing peptides, among other components. The CDM medium contains only a sugar, salts, amino acids and vitamins. The five enzymes are branched chain aminotransferase (BcaT), involved in the degradation of branched chain amino acids (ile, val, leu); alpha-hydroxyisocaproic acid dehydrogenase (HicD), involved in the degradation of aromatic amino acids; two peptidases PepN and PepXP, involved in the degradation of peptides to amino acids; finally, an esterase activity involved in the degradation of fatty acids. These enzymes were chosen because they are known to be involved in the development of flavour compounds in cheese.</p>
</div>
<div id="data-conversion" class="section level3">
<h3><span class="header-section-number">19.1.2</span> Data conversion</h3>
<ol style="list-style-type: decimal">
<li>Convert the “Cremoris.Like.Genotype” column into a logical vector instead of a numerical vector. Remember the fact that many functions in R, including the coercion functions themselves, have a default smart way of data coercion.
<div>
<a id="BiasMetabHead2" href="javascript:toggle('BiasMetabSol2','BiasMetabHead2');" >Show solution</a>
</div>
<div id="BiasMetabSol2" style="display: none">
<p>Using the <code>as.logical()</code> coercion function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d<span class="op">$</span>Cremoris.Like.Genotype &lt;-<span class="st"> </span><span class="kw">as.logical</span>(d<span class="op">$</span>Cremoris.Like.Genotype)</code></pre></div>
<p>Or if you don’t want to rely on the default coercion of <code>as.logical()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d<span class="op">$</span>Cremoris.Like.Genotype &lt;-<span class="st"> </span><span class="kw">ifelse</span>(d<span class="op">$</span>Cremoris.Like.Genotype<span class="op">==</span><span class="dv">1</span>,<span class="ot">TRUE</span>,<span class="ot">FALSE</span>)</code></pre></div>
</div></li>
<li>Make two index vectors that contain the column indices of the cultures grown on GM17 on the one hand and the cultures grown on CDM on the other hand.
<div>
<a id="BiasMetabHead3" href="javascript:toggle('BiasMetabSol3','BiasMetabHead3');" >Show solution</a>
</div>
<div id="BiasMetabSol3" style="display: none">
<p>We define <code>GM17</code> and <code>CDM</code> as the ranges:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GM17 &lt;-<span class="st"> </span><span class="dv">8</span><span class="op">:</span><span class="dv">12</span>
CDM &lt;-<span class="st"> </span><span class="dv">13</span><span class="op">:</span><span class="dv">17</span></code></pre></div>
</div></li>
<li>Using <code>apply()</code> or <code>lapply()</code> (remember, the <strong>data frame</strong> class is a subclass of the <strong>list</strong> class, so you can use <code>lapply()</code>), print the classes of the elements of the data frame that contain the enzyme activity data.
<div>
<a id="BiasMetabHead4" href="javascript:toggle('BiasMetabSol4','BiasMetabHead4');" >Show solution</a>
</div>
<div id="BiasMetabSol4" style="display: none">
<p>We use <code>apply()</code> to the columns:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">apply</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dv">2</span>, class)</code></pre></div>
<pre><code>##     GM17.BcaT     GM17.HicD     GM17.PepN    GM17.PepXP GM17.Esterase 
##   &quot;character&quot;   &quot;character&quot;   &quot;character&quot;   &quot;character&quot;   &quot;character&quot; 
##      CDM.BcaT      CDM.HicD      CDM.PepN     CDM.PepXP  CDM.Esterase 
##   &quot;character&quot;   &quot;character&quot;   &quot;character&quot;   &quot;character&quot;   &quot;character&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#lapply(d[,c(GM17,CDM)], class)</span></code></pre></div>
</div></li>
<li>Among the enzyme activity columns, convert those that are character class vectors to numeric vectors using <code>apply()</code> and the assignment operator. Check the classes of the data frame elements again.
<div>
<a id="BiasMetabHead5" href="javascript:toggle('BiasMetabSol5','BiasMetabHead5');" >Show solution</a>
</div>
<div id="BiasMetabSol5" style="display: none">
<p>Converting and checking the classes of the <code>GM17</code> and <code>CDM</code> ranges:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d[,<span class="kw">c</span>(GM17,CDM)] &lt;-<span class="st"> </span><span class="kw">apply</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dv">2</span>, as.numeric)
<span class="kw">apply</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dv">2</span>, class)</code></pre></div>
<pre><code>##     GM17.BcaT     GM17.HicD     GM17.PepN    GM17.PepXP GM17.Esterase 
##     &quot;numeric&quot;     &quot;numeric&quot;     &quot;numeric&quot;     &quot;numeric&quot;     &quot;numeric&quot; 
##      CDM.BcaT      CDM.HicD      CDM.PepN     CDM.PepXP  CDM.Esterase 
##     &quot;numeric&quot;     &quot;numeric&quot;     &quot;numeric&quot;     &quot;numeric&quot;     &quot;numeric&quot;</code></pre>
</div></li>
</ol>
</div>
<div id="data-analysis" class="section level3">
<h3><span class="header-section-number">19.1.3</span> Data analysis</h3>
<p>We will now explore the information present in the data. We are interested in differences in enzyme activities that we can link to the other characteristics of the strains. We are also interested in correlations between the activities, for example in cells grown in GM17 and in CDM.</p>
<ol style="list-style-type: decimal">
<li>Make a pairplot of all the enzyme activity columns. Use the ranges defined above to select the columns, and remember that the default plot method of a data frame is a pairplot of all columns against each other. Use a plotting symbol that gives a nice graph. Since the activity range between the different strains and conditions is considerable, make the same plot using logarithmically transformed activities. This often yields a better spread of the data in a plot.
<div>
<a id="BiasMetabHead6" href="javascript:toggle('BiasMetabSol6','BiasMetabHead6');" >Show solution</a>
</div>
<div id="BiasMetabSol6" style="display: none">
<p>Making two pairplots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dt">pch=</span><span class="dv">20</span>)
<span class="kw">plot</span>(<span class="kw">log</span>(d[,<span class="kw">c</span>(GM17,CDM)]), <span class="dt">pch=</span><span class="dv">20</span>)</code></pre></div>
</div></li>
<li>Calculate a correlation matrix of the enzyme activity data, using both Pearson and Spearman rank correlation definitions. There is one column causing difficulties because of missing values. Still, we want to incorporate it. Read the help file of the correlation function to learn how to do that.
<div>
<a id="BiasMetabHead7" href="javascript:toggle('BiasMetabSol7','BiasMetabHead7');" >Show solution</a>
</div>
<div id="BiasMetabSol7" style="display: none">
<p>Calculating the Pearson and Spearman correlation coefficients with <code>cor()</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cor</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dt">method=</span><span class="st">&#39;pearson&#39;</span>, <span class="dt">use=</span><span class="st">&#39;complete.obs&#39;</span>)
<span class="kw">cor</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dt">method=</span><span class="st">&#39;spearman&#39;</span>, <span class="dt">use=</span><span class="st">&#39;complete.obs&#39;</span>)</code></pre></div>
</div></li>
<li>We are mostly interested in positively or negatively correlated activities. Pairplots and correlation coefficients yield this information, but another efficient way to summarize these is a hierarchical tree, constructed using some measure of distance between the samples. As a distance measure, we could take (1 - correlation coefficient). There is a standard clustering function in R called <code>hclust()</code>. Find out what it needs as an input, and make a plot of the calculated tree object using its <code>plot()</code> method. What is the effect of different tree agglomeration methods and of different correlation coefficients (Pearson, Spearman)? Compare the trees with the pairplots.
<div>
<a id="BiasMetabHead8" href="javascript:toggle('BiasMetabSol8','BiasMetabHead8');" >Show solution</a>
</div>
<div id="BiasMetabSol8" style="display: none">
<p>Using Ward’s agglomeration method and Spearman’s rank correlation as the distance metric. The cluster tree nicely summarizes the information from the matrix of correlation coefficients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">distance &lt;-<span class="st"> </span><span class="kw">as.dist</span>(<span class="dv">1</span><span class="op">-</span><span class="kw">cor</span>(d[,<span class="kw">c</span>(GM17,CDM)], <span class="dt">method=</span><span class="st">&#39;spearman&#39;</span>, <span class="dt">use=</span><span class="st">&#39;complete.obs&#39;</span>))
hc &lt;-<span class="st"> </span><span class="kw">hclust</span>(distance, <span class="dt">method=</span><span class="st">&#39;ward.D&#39;</span>)
<span class="kw">plot</span>(hc, <span class="dt">hang=</span><span class="op">-</span><span class="dv">1</span>)</code></pre></div>
<p><img src="StatR_files/figure-html/classdairy_09-1.png" width="672" /></p>
</div></li>
<li>We would also like to investigate relations between activities and the type of strains. To explore these putative relations make pairplots, using your preferred data transformation, in which you give the points different colors or symbols, depending on the type of bacteria which was contained in the cheeses. To help you get started with this:
<ul>
<li>Create a vector of colors or symbol numbers having the same length as the number of levels in the factor that you want to use to distinguish points. For example, the number of levels in the “Variant” column equals 3. So, create a vector <code>mycolors &lt;- c('red','green','blue')</code>. These color names will be recognized by R.</li>
<li>Then, in the plot command use the argument <code>col=mycolors[d$Variant]</code>. The “<code>[</code>” function will automatically coerce the factor to a numeric vector, containing numbers 1, 2, or 3 (check <code>as.numeric(d$Variant)</code>). This numeric vector is then used as an index vector to select the appropriate color from the <code>mycolors</code> vector.
<div>
<a id="BiasMetabHead9" href="javascript:toggle('BiasMetabSol9','BiasMetabHead9');" >Show solution</a>
</div>
<div id="BiasMetabSol9" style="display: none">

<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="kw">log</span>(d[,<span class="kw">c</span>(GM17,CDM)]), <span class="dt">cex=</span><span class="fl">0.5</span>, <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&#39;red&#39;</span>,<span class="st">&#39;green&#39;</span>,<span class="st">&#39;blue&#39;</span>)[d<span class="op">$</span>Variant])
<span class="kw">plot</span>(<span class="kw">log</span>(d[,<span class="kw">c</span>(GM17,CDM)]), <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&#39;red&#39;</span>,<span class="st">&#39;green&#39;</span>)[d<span class="op">$</span>Origin])
<span class="kw">plot</span>(<span class="kw">log</span>(d[,<span class="kw">c</span>(GM17,CDM)]), <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&#39;red&#39;</span>,<span class="st">&#39;green&#39;</span>)[d<span class="op">$</span>Cremoris.Like.Genotype<span class="op">+</span><span class="dv">1</span>])</code></pre></div>
In the last command the function <code>+()</code> takes care of the coercion of the logical vector <code>d$Cremoris.Like.Genotype</code> to a numeric vector.</li>
</ul>
</div></li>
</ol>
<p>The previous plots show that colors (or symbol types) often cluster, suggesting that we could use the measured activities to distinguish strains. To explore this possibility we are going to use decision trees. If you don’t know what these are, look it up in the internet.</p>
<ol style="list-style-type: decimal">
<li><p>Install the R-package called <code>party</code>. Take a look at the help page of its <code>ctree()</code> function, in particular at the example that uses <code>ctree()</code> for the classification of the iris data. Also notice the special formula notation <code>Species ~ .</code>, where “.” stands for all variables in the data frame except those in the left hand side (“Species” here).</p></li>
<li>Make and plot a classification tree for the Origin (dairy vs non-dairy) of the strain. Also make one for the Variant. You will have to supply the <code>ctree()</code> function with customized data frames consisting of selected columns from the original data frame to be able to use the same formula syntax as in the iris example. Which factor seems to be highly specific for the dairy origin?
<div>
<a id="BiasMetabHead10" href="javascript:toggle('BiasMetabSol10','BiasMetabHead10');" >Show solution</a>
</div>
<div id="BiasMetabSol10" style="display: none">
<p>Using the “party” package we make decision trees that are able to distinguish strains of dairy- and non-dairy origin or strain variants:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(party)
origtree &lt;-<span class="st"> </span><span class="kw">ctree</span>(Origin<span class="op">~</span>., <span class="dt">data=</span>d[,<span class="kw">c</span>(<span class="dv">4</span>,GM17,CDM)])
<span class="kw">plot</span>(origtree)
vartree &lt;-<span class="st"> </span><span class="kw">ctree</span>(Variant<span class="op">~</span>., <span class="dt">data=</span>d[,<span class="kw">c</span>(<span class="dv">4</span>,<span class="dv">6</span>,GM17,CDM)])
<span class="kw">plot</span>(vartree)</code></pre></div>
<p>The PepXP enzyme activity is a strong determinant of the origin of strains, where dairy origin seems to be associated with high activity of this enzyme. This is not surprising, because the PepXP peptidase is involved in the degradation of milk protein to amino acids. Amino acids are needed by these bacteria to be able to grow.</p>
</div></li>
<li>Compare prediction and actual values for these classification trees: note that <code>ctree</code> objects also have a <code>predict()</code> method!
<div>
<a id="BiasMetabHead11" href="javascript:toggle('BiasMetabSol11','BiasMetabHead11');" >Show solution</a>
</div>
<div id="BiasMetabSol11" style="display: none">
<p>To compare actual and predicted classes we can make a list like:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cbind</span>(<span class="kw">as.character</span>(d<span class="op">$</span>Origin),<span class="kw">as.character</span>(<span class="kw">predict</span>(origtree)))</code></pre></div>
</div></li>
<li>Can you give a critique concerning the structure of the current data set, when trying to make a decision tree differentiating the origin of the strains? Support your critique with numbers!
<div>
<a id="BiasMetabHead12" href="javascript:toggle('BiasMetabSol12','BiasMetabHead12');" >Show solution</a>
</div>
<div id="BiasMetabSol12" style="display: none">
<p>Critique: the data is highly unbalanced with respect to origin and variant:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(d<span class="op">$</span>Origin)</code></pre></div>
<pre><code>## 
##     dairy non-dairy 
##        25        59</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(d<span class="op">$</span>Variant)</code></pre></div>
<pre><code>## 
##      cremoris diacetylactis        lactis 
##             9             7            68</code></pre>
<p>This leads to bias in the selection of variables used for the tree. Trees constructed using such data will probably not be good predictors for strains from unknown class. Furthermore, there is a high correlation between origin and variant types. Almost all lactis types are of non-dairy origin, all cremoris types are of dairy origin and almost all diacetylactis types are of dairy origin:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(<span class="kw">interaction</span>(d<span class="op">$</span>Origin,d<span class="op">$</span>Variant))</code></pre></div>
<pre><code>## 
##          dairy.cremoris      non-dairy.cremoris     dairy.diacetylactis 
##                       9                       0                       6 
## non-dairy.diacetylactis            dairy.lactis        non-dairy.lactis 
##                       1                      10                      58</code></pre>
</div></li>
<li>How could you test the accuracy of such a prediction tree with the current data set?
<div>
<a id="BiasMetabHead13" href="javascript:toggle('BiasMetabSol13','BiasMetabHead13');" >Show solution</a>
</div>
<div id="BiasMetabSol13" style="display: none">
<p>The prediction trees could, for example, be tested with the leave-one-out procedure. Take one row out of the data set and construct a tree predictor from the remaining data. Then predict the class of the excluded row, and note whether it is correct or not. Do this for all rows. This gives an impression of the accuracy. However, the unbalancedness of the data will remain a problem.</p>
</div></li>
</ol>

</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Bachmann2009">
<p>Bachmann, H., M. J. C. Starrenburg, A. Dijkstra, D. Molenaar, M. Kleerebezem, J. L. W. Rademaker, and J. E. T. Van Hylckama Vlieg. 2009. “Regulatory Phenotyping Reveals Important Diversity Within the Species <em>Lactococcus Lactis</em>.” <em>Appl. Environ. Microbiol.</em> 75 (17): 5687–94. doi:<a href="https://doi.org/10.1128/AEM.00919-09">10.1128/AEM.00919-09</a>.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="14">
<li id="fn14"><p>The sequence of the small subunit ribosomal RNA (SSU-rRNA)<a href="class-dairy-strains.html#fnref14">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="classifiers.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="naivebayes.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": {}
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"download": null,
"toc": {
"collapse": "subsection",
"scroll_highlight": true
},
"toolbar": {
"position": "fixed"
},
"search": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "libs/mathjax-local/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
